crates/superposition_derives/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::punctuated::Punctuated;
+use syn::{parse::Parse, parse_macro_input, DeriveInput, Error, LitStr, Meta, Token};
+
+#[proc_macro_derive(CustomQuery, attributes(custom_query))]
+pub fn derive_custom_query(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = &input.ident;
+    let generics = &input.generics;
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let custom_query_attr = match extract_attributes(&input.attrs) {
+        Ok(Some(attr)) => attr,
+        Ok(None) => {
+            return Error::new_spanned(&input, "Missing #[custom_query] attribute")
+                .to_compile_error()
+                .into()
+        }
+        Err(e) => return e.to_compile_error().into(),
+    };
+
+    let regex_pattern = &custom_query_attr.regex;
+    let capture_group = &custom_query_attr.capture_group;
+
+    let inner_type = generics
+        .type_params()
+        .next()
+        .map(|t| t.ident.clone())
+        .unwrap_or_else(|| syn::Ident::new("T", proc_macro2::Span::call_site()));
+
+    let expanded = quote! {
+        impl #impl_generics superposition_types::custom_query::CustomQuery for #name #ty_generics #where_clause {
+            type Inner = #inner_type;
+
+            fn regex_pattern() -> &'static str {
+                #regex_pattern
+            }
+
+            fn capture_group() -> &'static str {
+                #capture_group
+            }
+
+            fn from_inner(inner: Self::Inner) -> Self {
+                Self(inner)
+            }
+        }
+
+        impl #impl_generics #name #ty_generics #where_clause {
+            pub fn into_inner(self) -> #inner_type {
+                self.0
+            }
+        }
+
+        impl #impl_generics actix_web::FromRequest for #name #ty_generics #where_clause {
```

Comment:
shouldn't this be left on individual types to implement?
here we will be able to return `HashMap<String, String>` and the dimensions sent needs to converted in `Map<String, Value>` in most of our use cases